mbedtls/src/ssl/context.rs

```diff
@@ -172,35 +172,118 @@ impl<T> Context<T> {
+    ssl_set_bio(
+        ctx,
+        io as *mut T as *mut c_void,
+        Some(T::call_send),
+        Some(T::call_recv),
+        None,
+    );
+}
+
+/// This function provides a way to apply async context to bio before running 
+/// any logic.
+/// Note: `bio` is a concept in common TLS implementation which refers to basic IO.
+/// openssl and mbedtls both use this concept.
+/// Ref: https://stackoverflow.com/questions/51672133/what-are-openssl-bios-how-do-they-work-how-are-bios-used-in-openssl
+#[cfg(all(feature = "std", feature = "async"))]
+impl<T> Context<T>  {
+    pub(super) fn with_bio_async<'cx, R, IoType>(&mut self, cx: &mut std::task::Context<'cx>, f: impl FnOnce(&mut Self) -> R) -> Option<R> where for<'c> (&'c mut std::task::Context<'cx>, &'c mut T): IoCallbackUnsafe<IoType> {
+        let ret;
+
+        struct BioGuard<'a, T> {
+            context: &'a mut Context<T>,
+        }
+        
+        impl<'a, T> Drop for BioGuard<'a, T> {
+            fn drop(&mut self) {
+                self.context.clear_bio();
+            }
+        }
+        // SAFETY: In the call to `set_bio_raw`, `user_data` must live as long
+        // as `ctx`, or until the bio is cleared from `ctx`. The bio is cleared
+        // at the end of this block ensured by the drop guard: [`BioGuard`]
+        unsafe {
+            // Points to `self.inner`, so safe to borrow at the same time as `self.io`
+            let ctx = self.into();
+            let mut user_data = (cx, &mut**self.io.as_mut()?);
+            set_bio_raw(ctx, &mut user_data);
+
+            let guard = BioGuard { context: self };
+
+            ret = f(guard.context);
+        }
+
+        Some(ret)
+    }
+
+    // This function is created to handle the old behavior of  `mbedtls_ssl_write()` only for TLS use cases
+    // If there is pending data prepared in an earlier call to `mbedtls_ssl_write()` in c-mbedtls's buffer, this function try to flush them first by calling `mbedtls_ssl_flush_output`.
+    // Then call `mbedtls_ssl_write()` as normal if all pending data has been flushed out.
+    // It handles error as normal because in case of getting `SslWantWrite` from `mbedtls_ssl_write()` or `mbedtls_ssl_flush_output()`, this function just need to be called again.
+    // Ref: https://github.com/Mbed-TLS/mbedtls/issues/4183
+    pub(super) fn async_write(&mut self, buf: &[u8]) -> Result<usize> {
+        unsafe {
+            let inner: *mut ssl_context = self.into();
+            if (*inner).out_left > 0 {
+                match self.flush_output() {
+                    Ok(()) => {
+                        // if data prepared in an earlier call to `mbedtls_ssl_write()` is all flushed then continue to write new data
+                        if (*inner).out_left != 0 {
+                            // data prepared in an earlier call has not been flushed out
+                            // return `SslWantWrite` to caller to indicate call this function again later
```

Comment:
I think we should keep looping until we get an error or out_left is 0. Or perhaps we need to interpret the positive return values of flush_output? In particular, we need to be sure that the inner IO has set up notifications for the current task.

Reply:
> I think we should keep looping until we get an error or out_left is 0
I could follow this approach.

Regarding:
> In particular, we need to be sure that the inner IO has set up notifications for the current task.
If just let Future's mechanism to do the retry, I understand that `cx.waker().clone().wake();` in `poll_write` already do the work for sending  notifications?